src/consensus/params.cpp

```diff
@@ -163,17 +164,26 @@ namespace Consensus {
 
         // Parse the address strings into concrete types.
         std::vector<FundingStreamAddress> addresses;
-        for (auto addr : strAddresses) {
-            auto taddr = keyIO.DecodeDestination(addr);
-            if (IsValidDestination(taddr)) {
-                addresses.push_back(GetScriptForDestination(taddr));
-            } else {
-                auto zaddr = keyIO.DecodePaymentAddress(addr);
-                // If the string is not a valid transparent or Sapling address, we will
-                // throw here.
-
-                addresses.push_back(std::get<libzcash::SaplingPaymentAddress>(zaddr));
+        for (const auto& strAddr : strAddresses) {
+            auto addr = keyIO.DecodePaymentAddress(strAddr);
+            if (!addr.has_value()) {
+                throw std::runtime_error("Funding stream address was not valid Zcash address.");
             }
+
+            std::visit(match {
+                [&](const CKeyID& keyId) {
+                    addresses.push_back(GetScriptForDestination(keyId));
+                },
```

Comment:
The spec says, in [section 7.10 'Payment of Funding Streams'](https://zips.z.cash/protocol/protocol.pdf#fundingstreams):
> Each element of fs.AddressList MUST represent either a transparent P2SH address as specified in §5.6.1.1 ‘Transparent Addresses’ on p. 108, or a Sapling shielded payment address as specified in §5.6.3.1 ‘Sapling Payment Addresses’ on p. 111.

I.e. it is invalid for a funding stream to go to a P2PKH address. I guess it doesn't matter too much because the funding stream addresses are hardcoded, but I would do this:

```suggestion
```

Reply:
If we interpret the protocol spec that way, then I think it's an overconstraint on the part of the protocol spec. I think the funding stream implementation in `zcashd` should enable funding stream outputs to go to whatever address kinds are supported as coinbase outputs, with the precise addresses being what is specified in consensus. That way, when we inevitably alter the funding stream addresses in a subsequent NU, the only thing we need to change is the list of addresses (and the protocol spec, which we'd do anyway).